runtime-sdk/src/dispatcher.rs

```diff
@@ -467,8 +447,148 @@ impl<R: Runtime + Send + Sync> transaction::dispatcher::Dispatcher for Dispatche
+                    let tx_size = tx.len().try_into().map_err(|_| {
+                        Error::MalformedTransactionInBatch(anyhow!("transaction too large"))
+                    })?;
+                    // It is an error to include a malformed transaction in a batch. So instead of only
+                    // reporting a failed execution result, we fail the whole batch. This will make the compute
+                    // node vote for failure and the round will fail.
+                    //
+                    // Correct proposers should only include transactions which have passed check_tx.
+                    let tx = Self::decode_tx(ctx, tx)
+                        .map_err(|err| Error::MalformedTransactionInBatch(err.into()))?;
+                    txs.push((tx_size, tx.clone()));
+
+                    if prefetch_enabled {
+                        Self::prefetch_tx(&mut prefixes, tx)?;
+                    }
+                }
+                if prefetch_enabled {
+                    ctx.runtime_state()
+                        .prefetch_prefixes(prefixes.into_iter().collect(), R::PREFETCH_LIMIT);
+                }
+
+                // Execute the batch.
+                let mut results = Vec::with_capacity(batch.len());
+                for (index, (tx_size, tx)) in txs.into_iter().enumerate() {
+                    results.push(Self::execute_tx(ctx, tx_size, tx, index)?);
+                }
+
+                Ok(results)
+            },
+        )
+    }
+
+    fn schedule_and_execute_batch(
+        &self,
+        rt_ctx: transaction::Context<'_>,
+        batch: &mut TxnBatch,
+    ) -> Result<ExecuteBatchResult, RuntimeError> {
+        let cfg = R::SCHEDULE_CONTROL.unwrap(); // Must succeed otherwise we wouldn't be here.
+        let mut tx_reject_hashes = Vec::new();
+
+        let mut result = self.execute_batch_common(
+            rt_ctx,
+            |ctx| -> Result<Vec<ExecuteTxResult>, RuntimeError> {
+                // Schedule and execute the batch.
+                //
+                // The idea is to keep scheduling transactions as long as we have some space
+                // available in the block as determined by gas use.
+                let mut new_batch = Vec::new();
+                let mut results = Vec::with_capacity(batch.len());
+                let mut requested_batch_len = cfg.initial_batch_size;
+                'batch: loop {
+                    // Remember length of last batch.
+                    let last_batch_len = batch.len();
+                    let last_batch_tx_hash = batch.last().map(|raw_tx| Hash::digest_bytes(raw_tx));
+
+                    for raw_tx in batch.drain(..) {
+                        // If we don't have enough gas for processing even the cheapest transaction
+                        // we are done.
+                        let remaining_gas = R::Core::remaining_batch_gas(ctx);
+                        if remaining_gas < cfg.min_remaining_gas {
```

Comment:
:information_source: this prevents us from decoding a tx when there's less gas than the cheapest possible transaction remaining.

any reason we might set this to a number other than the cheapest possible transaction, btw?

Reply:
Nope, this should be set to the gas cost of the cheapest possible transaction. I just didn't want to add an automated mechanism to determine that.